learn/introduction-to-scryer.dj

```diff
@@ -0,0 +1,319 @@
+Visual Studio Code with the Better Prolog Syntax extension can highlight Prolog code.  But currently (2025) there is no way to execute Scryer Prolog in Visual Studio Code.  You will have to use the terminal.
+
+## Testing Frameworks
+
+Only if you are experienced with testing frameworks and package managers for other languages, and think that having a testing framework available will help you learn, consider installing [bakage](https://github.com/bakaq/bakage) and then use bakage to install [testing.pl](https://github.com/bakaq/testing.pl).
+
+Otherwise, learn Prolog until you are somewhat comfortable with the system and concepts.  
+
+## Load a Prolog Module
+
+In the earlier example of hello world, the hello.pl Prolog module was 'consulted' from the
+current working directory.
+
+```bash
+$ scryer-prolog 
+?- [hello].
+Hello, world!
+   true.
+?- hello_prolog.
+Hello, Prolog world!
+   true.
+?- 
+```
+
+There are more ways to consult a Prolog module, and much more you can do in the toplevel.
+
+The toplevel is explained on [https://www.metalevel.at/prolog/concepts](https://www.metalevel.at/prolog/concepts).
+
+
+
+## Errors and Warnings
+
+### Instantiation Error
+
+Above we recommended staying away from (is)/2, for now.  And that recommendation still holds.  It is useful, though, to identify and understand some of the errors you will experience in Scryer Prolog (and most other Prolog systems).
+
+
+```prolog
+?- A is 3.
+   A = 3.
+?- 3 is  A.
+   error(instantiation_error,(is)/2).
+?- 
+```
+
+```prolog
+?- X is "hello".
+   error(type_error(evaluable,[]/0),(is)/2).
+?- 
+
+```
+
+clpz:(#=) is more friendly.
+
+```prolog
+?- 3 #= X.
+   X = 3.
+?- 
+```
+
+You will eventually run into other instantiation_errors. Hopefully you
+will have learned a little Prolog by then and will be able to solve them.
+
+### Discontiguous Clause Warnings
+
+Pay attention to these warnings. It can be hard to identify the cause, but the warning likely indicates a problem.
+
+```prolog
+?- [user].
+ce(a).
+ce(b).
+ce(c).
+ce(d).
+cd(e).
+ce(f).
+ce(g).
+ce(h).
+
+% Warning: overwriting ce/1 because the clauses are discontiguous
+?- 
+
+```
+If you can't eyeball where the error is, introduce a syntax error 
+in your program and move it around until you locate the warning.
+
+
+### : instead of :-
+
+Writing : instead of :- can result in difficult-to-find 
+ errors, even when you know the cause:
+
+colon.pl:
+
+```prolog
+a(-1) :- true.
+a(3) :-  3#>4.
+a(2) : true.
+a(_) :-  X#>12.
+```
+
+```prolog
+?- [colon].
+   error(permission_error(modify,static_procedure,(:)/2),load/1).
+?- 
+
+```
+
+In a longer file, you may not be able to detect the missing - visually.
+You can try deliberately introducing a syntax error in your Prolog program to help locate the line where the error occurs.  Keep adjusting the location
+of the deliberate syntax error until you find the line where the permission_error occurs.
```

Comment:
Would it not be better to have Scryer report the line where the error occurs? This is ideally not something that must be done manually.

Reply:
Yes.  If Scryer can do that I don't know how.

Reply:
added a comment about no line number reporting .